text/0000-otlp-trace-data-format.md

```diff
@@ -0,0 +1,361 @@
+# OTLP Trace Data Format
+
+_Author: Tigran Najaryan, Splunk_
+
+**Status:** `proposed`
+
+OTLP Trace Data Format specification describes the structure of the trace data that is transported by OpenTelemetry Protocol (RFC0035).
+
+## Motivation
+
+This document is a continuation of OpenTelemetry Protocol RFC0035 and is necessary part of OTLP specification.
+
+## Explanation
+
+OTLP Trace Data Format is primarily inherited from OpenCensus protocol. Several changes are introduced with the goal of more efficient serialization. Notable differences from OpenCensus protocol are:
+
+1. Removed `Node` as a concept.
+2. Extended `Resource` to better describe the source of the telemetry data.
+3. Replaced attribute maps by lists of key/value pairs.
+4. Eliminated unnecessary additional nesting in various values.
+
+Changes 1-2 are conceptual, changes 3-4 improve performance.
+
+## Internal details
+
+This section specifies data format in Protocol Buffers.
+
+### Resource
+
+```
+// Resource information. This describes the source of telemetry data.
+message Resource {
+  // Set of labels that describe the resource. See OpenTelemetry specification
+  // semantic conventions for standardized label names:
+  // https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md
+
+  repeated AttributeKeyValue labels = 3;
+  int32 dropped_labels_count = 11;
+}
+```
+
+### Span
+
+```
+// A span represents a single operation within a trace. Spans can be
+// nested to form a trace tree. Spans may also be linked to other spans
+// from the same or different trace. And form graphs. Often, a trace
+// contains a root span that describes the end-to-end latency, and one
+// or more subspans for its sub-operations. A trace can also contain
+// multiple root spans, or none at all. Spans do not need to be
+// contiguous - there may be gaps or overlaps between spans in a trace.
+//
+// The next id is 17.
+message Span {
+  // A unique identifier for a trace. All spans from the same trace share
+  // the same `trace_id`. The ID is a 16-byte array. An ID with all zeroes
+  // is considered invalid.
+  //
+  // This field is semantically required. Receiver should generate new
+  // random trace_id if empty or invalid trace_id was received.
```

Comment:
I think they should reject the message on invalid trace_id

Reply:
@yurishkuro I would argue that what you suggest is against robustness principle: https://en.wikipedia.org/wiki/Robustness_principle of which I am a fan and which has its share of critics. I would like to hear your thoughts. :-)

Reply:
Speaking of SD principles to be a fan of, here's another one: fail fast. If the first receiver consistently rejects malformed messages, the sender has a better chance to figure out what they are doing wrong than if they have to dig up that message several layers down.

An invalid trace id means that even if we store the span, we won't be able to correlate it with the rest of the trace, making the data nearly useless.

There is one reason to store even the bad data, to be able to track it back to the source (bottom-up) and file a ticket against the team that generates it. The bad data needs to be clearly labeled in this case. For example, in Jaeger spans we have a section "warnings", and when we perform any data overrides we make a note in that list (with the original value).

In the end, I don't think it's a question for the data format spec, I'd expect this behavior to be configurable in the receiver: rejects, accept and clean up, or do 50/50 (to both warn the client and keep the data for bottom-up troubleshooting). But it's fine to have a recommendation in the data spec on *how* to do it (as you did).

Reply:
I tend to agree with you. I'd say robustness principle would apply more if there was a good way to substitute bad data with good data or interpret it in a way that makes most sense.
Replacing missing trace id with a new random id is probably not the best approach because we will silently lose the chance to detect the problem early.

I would say a likely better approach is to leave the field as is (empty or invalid), record the fact (e.g. in a telemetry counter or wherever else it makes sense) and in addition to that for example:

- If the use case is to forward the span (e.g. an Agent or Collector) then forward it as is.
- If however the use case requires specific interpretation of traceid (for example head sampling based on a hash of a traceid) then this should be considered bad input data and entire Span should be dropped/recorded as bad/etc.

Because the best possible course of action can be different for each use case I think it is best to remove the current recommendation to generate a new traceid and instead recommend to:

"The code that receives such Span SHOULD record the fact of invalid or missing traceid. If valid traceid and an interpretation of it is necessary for the use case then Span SHOULD be considered to be a case of bad input data otherwise the invalid or missing traceid MAY be ignored".

Thoughts?

Reply:
I still think being too rigid about what "must happen" is overkill at this stage. I would instead propose possible behaviors, but not dictate them:
  * the receiver MAY reject the invalid data and respond with the appropriate error code to the sender
  * the receiver MAY accept the invalid data and attempt to correct it. In this case the receiver SHOULD "log" the corrections it made (record where ???)

For the last part, I think it would be useful to provide a slot in the span model for these "logs" (as in https://github.com/jaegertracing/jaeger/blob/master/model/proto/model.proto#L136)

Reply:
"MAY" sounds reasonable to me. I will add them as possible recommendations. I will drop the "log" part since there are many ways to do it (log locally, record a metric, log in the attributes, etc) and it is hard to recommend single best universal way.

Reply:
Fixed.